File: Source/counseling.py
"""
Genetic Counseling Referral System for Mergenix.

Determines when genetic counseling is recommended based on analysis results,
generates tier-gated referral summaries, and provides a searchable counselor
directory filtered by specialty and location.
"""

from __future__ import annotations

import os
from datetime import date

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
NSGC_URL = "https://www.nsgc.org/findageneticcounselor"

COUNSELING_SPECIALTIES = [
    "prenatal",
    "carrier_screening",
    "cancer",
    "cardiovascular",
    "pediatric",
    "neurogenetics",
    "pharmacogenomics",
    "general",
]

_DATA_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data")
_PROVIDERS_PATH = os.path.join(_DATA_DIR, "counseling_providers.json")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def get_counseling_specialties() -> list[str]:
    """Return the list of recognized genetic counseling specialties."""
    return list(COUNSELING_SPECIALTIES)


def should_recommend_counseling(
    carrier_results: list[dict],
    prs_results: list[dict] | None = None,
    pgx_results: list[dict] | None = None,
) -> tuple[bool, list[str]]:
    """Decide whether genetic counseling should be recommended.

    Args:
        carrier_results: Output of ``analyze_carrier_risk`` — each dict has
            at minimum ``risk_level``, ``condition``, ``parent_a_status``,
            ``parent_b_status``.
        prs_results: Optional polygenic risk score results.  Each dict should
            contain ``percentile`` (float 0-100) and ``trait`` (str).
        pgx_results: Optional pharmacogenomics results.  Each dict should
            contain ``actionable`` (bool) and ``drug`` (str).

    Returns:
        A ``(recommend, reasons)`` tuple where *recommend* is ``True`` when
        counseling is advised and *reasons* lists human-readable explanations.
    """
    reasons: list[str] = []

    # 1. Both parents are carriers for the same disease
    for r in carrier_results:
        if r.get("parent_a_status") == "carrier" and r.get("parent_b_status") == "carrier":
            reasons.append(
                f"Both parents are carriers for {r['condition']} — "
                f"offspring have a 25% chance of being affected"
            )

    # 2. Any high-risk result
    for r in carrier_results:
        if r.get("risk_level") == "high_risk":
            msg = f"High-risk result detected for {r['condition']}"
            if msg not in reasons:
                reasons.append(msg)

    # 3. High PRS percentile (>90th)
    if prs_results:
        for p in prs_results:
            if p.get("percentile", 0) > 90:
                reasons.append(
                    f"Polygenic risk score for {p.get('trait', 'unknown trait')} "
                    f"is in the {p['percentile']:.0f}th percentile (elevated)"
                )

    # 4. Actionable PGx findings
    if pgx_results:
        for g in pgx_results:
            if g.get("actionable"):
                reasons.append(
                    f"Actionable pharmacogenomic finding for {g.get('drug', 'unknown drug')}"
                )

    return (len(reasons) > 0, reasons)


def generate_referral_summary(
    carrier_results: list[dict],
    user_name: str = "",
    tier: str = "free",
) -> dict:
    """Create a tier-gated referral summary from carrier analysis results.

    Args:
        carrier_results: Output of ``analyze_carrier_risk``.
        user_name: Optional display name for the referral letter.
        tier: One of ``"free"``, ``"premium"``, ``"pro"`` (case-insensitive).

    Returns:
        A dict with keys:
        - ``recommend`` (bool)
        - ``reasons`` (list[str])
        - ``nsgc_url`` (str) — always present
        - ``summary_text`` (str | None) — ``None`` for free tier
        - ``key_findings`` (list[dict] | None) — ``None`` for free tier
        - ``recommended_specialties`` (list[str] | None) — ``None`` for free tier
        - ``referral_letter`` (str | None) — only for pro tier
    """
    tier_lower = tier.lower()
    recommend, reasons = should_recommend_counseling(carrier_results)

    base: dict = {
        "recommend": recommend,
        "reasons": reasons,
        "nsgc_url": NSGC_URL,
        "summary_text": None,
        "key_findings": None,
        "recommended_specialties": None,
        "referral_letter": None,
    }

    if tier_lower == "free":
        return base

    # Premium & Pro: full summary ------------------------------------------------
    key_findings = _extract_key_findings(carrier_results)
    specialties = _infer_specialties(carrier_results)

    high_risk_count = sum(1 for r in carrier_results if r.get("risk_level") == "high_risk")
    carrier_count = sum(1 for r in carrier_results if r.get("risk_level") == "carrier_detected")

    summary_text = (
        f"Mergenix Genetic Counseling Summary\n"
        f"Generated: {date.today().isoformat()}\n\n"
        f"Diseases analyzed: {len(carrier_results)}\n"
        f"High-risk findings: {high_risk_count}\n"
        f"Carrier findings: {carrier_count}\n\n"
    )
    if recommend:
        summary_text += "Recommendation: Genetic counseling is advised.\n"
        for i, reason in enumerate(reasons, 1):
            summary_text += f"  {i}. {reason}\n"
    else:
        summary_text += "Recommendation: No urgent findings. Routine counseling optional.\n"

    base["summary_text"] = summary_text
    base["key_findings"] = key_findings
    base["recommended_specialties"] = specialties

    # Pro: referral letter -------------------------------------------------------
    if tier_lower == "pro":
        base["referral_letter"] = _format_referral_letter(
            user_name=user_name,
            key_findings=key_findings,
            specialties=specialties,
            reasons=reasons,
        )

    return base


def find_providers_by_specialty(
    providers: list[dict],
    specialty: str | None = None,
    state: str | None = None,
) -> list[dict]:
    """Filter a provider list by specialty and/or US state.

    Args:
        providers: List of provider dicts (see ``counseling_providers.json``).
        specialty: If given, only return providers whose ``specialty`` list
            contains this value (case-insensitive).
        state: If given, only return providers in this US state (2-letter code,
            case-insensitive).

    Returns:
        Filtered list of provider dicts.
    """
    results = providers

    if specialty:
        spec_lower = specialty.lower()
        results = [
            p for p in results
            if spec_lower in [s.lower() for s in p.get("specialty", [])]
        ]

    if state:
        state_upper = state.upper()
        results = [
            p for p in results
            if p.get("state", "").upper() == state_upper
        ]

    return results


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _extract_key_findings(carrier_results: list[dict]) -> list[dict]:
    """Pull out the most clinically relevant findings for the summary."""
    findings: list[dict] = []
    for r in carrier_results:
        risk = r.get("risk_level", "low_risk")
        if risk in ("high_risk", "carrier_detected"):
            findings.append({
                "condition": r["condition"],
                "gene": r.get("gene", ""),
                "risk_level": risk,
                "parent_a_status": r.get("parent_a_status", "unknown"),
                "parent_b_status": r.get("parent_b_status", "unknown"),
                "inheritance": r.get("inheritance", "autosomal_recessive"),
            })
    return findings


def _infer_specialties(carrier_results: list[dict]) -> list[str]:
    """Suggest counseling specialties based on the disease categories found."""
    specs: set[str] = set()

    for r in carrier_results:
        risk = r.get("risk_level", "low_risk")
        if risk not in ("high_risk", "carrier_detected"):
            continue

        category = r.get("category", "").lower()

        # Map disease categories to counseling specialties
        if "cancer" in category or "oncolog" in category:
            specs.add("cancer")
        if "cardio" in category or "heart" in category:
            specs.add("cardiovascular")
        if "neuro" in category or "brain" in category:
            specs.add("neurogenetics")
        if "pediatr" in category:
            specs.add("pediatric")

        # Carrier screening is always relevant for carrier results
        if risk == "carrier_detected" or (
            r.get("parent_a_status") == "carrier"
            and r.get("parent_b_status") == "carrier"
        ):
            specs.add("carrier_screening")

        # Prenatal is recommended for any high-risk finding
        if risk == "high_risk":
            specs.add("prenatal")

    # Always include general if we have any findings
    if specs:
        specs.add("general")

    return sorted(specs)


def _format_referral_letter(
    user_name: str,
    key_findings: list[dict],
    specialties: list[str],
    reasons: list[str],
) -> str:
    """Generate a formatted referral letter for pro-tier users."""
    name_line = f"Patient: {user_name}\n" if user_name else ""
    today = date.today().isoformat()

    lines = [
        "GENETIC COUNSELING REFERRAL",
        "=" * 40,
        f"Date: {today}",
        name_line.rstrip(),
        "",
        "Dear Genetic Counselor,",
        "",
        "This referral is generated by the Mergenix genetic analysis platform.",
        "The following findings warrant professional genetic counseling:",
        "",
    ]

    if reasons:
        for i, reason in enumerate(reasons, 1):
            lines.append(f"  {i}. {reason}")
        lines.append("")

    if key_findings:
        lines.append("Key Findings:")
        for f in key_findings:
            lines.append(
                f"  - {f['condition']} ({f['gene']}): "
                f"Parent A={f['parent_a_status']}, Parent B={f['parent_b_status']} "
                f"[{f['inheritance']}]"
            )
        lines.append("")

    if specialties:
        lines.append(f"Recommended specialties: {', '.join(specialties)}")
        lines.append("")

    lines.extend([
        "This report is for informational purposes only and does not constitute",
        "a medical diagnosis. Please review the raw data and clinical context.",
        "",
        "Sincerely,",
        "Mergenix Genetic Analysis Platform",
        f"NSGC Directory: {NSGC_URL}",
    ])

    return "\n".join(lines)

File: packages/genetics-engine/src/counseling.ts
/**
 * Genetic Counseling Referral System
 *
 * Determines when genetic counseling is recommended based on analysis results,
 * generates tier-gated referral summaries, and provides a searchable counselor
 * directory filtered by specialty and location.
 *
 * Counseling recommendation triggers:
 * 1. Both parents are carriers for the same disease (25% offspring affected risk)
 * 2. Any high-risk result in carrier analysis
 * 3. High PRS percentile (>90th percentile)
 * 4. Actionable pharmacogenomic findings
 *
 * Tier gating:
 * - Free: recommendation + NSGC link only
 * - Premium: full summary with key findings and specialties
 * - Pro: everything + formatted referral letter
 *
 * Ported from Source/counseling.py (323 lines).
 */

import type {
  CarrierResult,
  CounselingResult,
  CounselingKeyFinding,
  CounselorSpecialty,
  CounselingUrgency,
  Tier,
  CounselingProviderEntry,
} from './types';

// ─── Constants ──────────────────────────────────────────────────────────────

/** NSGC Find a Genetic Counselor URL. */
const NSGC_URL = 'https://www.nsgc.org/findageneticcounselor';

// ─── Urgency Classification ─────────────────────────────────────────────────

/**
 * Compute counseling urgency from carrier analysis results.
 *
 * - 'high': Any high_risk carrier result OR both parents are carriers for the
 *   same disease (25% offspring affected risk).
 * - 'moderate': Any carrier_detected result.
 * - 'informational': No actionable findings.
 *
 * @param carrierResults - Output of analyzeCarrierRisk()
 * @returns CounselingUrgency level
 */
function computeUrgency(carrierResults: CarrierResult[]): CounselingUrgency {
  let hasCarrierDetected = false;

  for (const r of carrierResults) {
    // High urgency: any high_risk result
    if (r.riskLevel === 'high_risk') {
      return 'high';
    }
    // High urgency: both parents are carriers for the same disease
    if (r.parentAStatus === 'carrier' && r.parentBStatus === 'carrier') {
      return 'high';
    }
    if (r.riskLevel === 'carrier_detected') {
      hasCarrierDetected = true;
    }
  }

  if (hasCarrierDetected) {
    return 'moderate';
  }

  return 'informational';
}

// ─── Recommendation Logic ───────────────────────────────────────────────────

/**
 * Decide whether genetic counseling should be recommended.
 *
 * Evaluates multiple trigger conditions from carrier analysis, PRS, and PGx
 * results and returns a recommendation with specific reasons.
 *
 * Trigger conditions (from Source/counseling.py should_recommend_counseling):
 * 1. Both parents are carriers for the same disease
 * 2. Any high-risk carrier result
 * 3. PRS percentile > 90th for any condition
 * 4. Any actionable PGx finding
 *
 * Ported from Source/counseling.py `should_recommend_counseling()`.
 *
 * @param carrierResults - Output of analyzeCarrierRisk()
 * @param prsResults - Optional PRS results (array of {percentile, trait} objects)
 * @param pgxResults - Optional PGx results (array of {actionable, drug} objects)
 * @returns Tuple of [recommend, reasons]
 */
export function shouldRecommendCounseling(
  carrierResults: CarrierResult[],
  prsResults?: Array<{ percentile: number; trait: string }>,
  pgxResults?: Array<{ actionable: boolean; drug: string }>,
): [boolean, string[]] {
  const reasons: string[] = [];

  // 1. Both parents are carriers for the same disease
  for (const r of carrierResults) {
    if (r.parentAStatus === 'carrier' && r.parentBStatus === 'carrier') {
      let riskDescription: string;
      if ('sons' in r.offspringRisk) {
        const sons = r.offspringRisk.sons.affected;
        const daughters = r.offspringRisk.daughters.affected;
        riskDescription = `${sons}% of sons and ${daughters}% of daughters may be affected`;
      } else {
        riskDescription = `offspring have a ${r.offspringRisk.affected}% chance of being affected`;
      }
      reasons.push(`Both parents are carriers for ${r.condition} \u2014 ${riskDescription}`);
    }
  }

  // 2. Any high-risk result
  for (const r of carrierResults) {
    if (r.riskLevel === 'high_risk') {
      const msg = `High-risk result detected for ${r.condition}`;
      if (!reasons.includes(msg)) {
        reasons.push(msg);
      }
    }
  }

  // 3. High PRS percentile (>90th)
  if (prsResults) {
    for (const p of prsResults) {
      if (p.percentile > 90) {
        reasons.push(
          `Polygenic risk score for ${p.trait} ` +
            `is in the ${Math.round(p.percentile)}th percentile (elevated)`,
        );
      }
    }
  }

  // 4. Actionable PGx findings
  if (pgxResults) {
    for (const g of pgxResults) {
      if (g.actionable) {
        reasons.push(`Actionable pharmacogenomic finding for ${g.drug}`);
      }
    }
  }

  return [reasons.length > 0, reasons];
}

/**
 * Create a tier-gated referral summary from carrier analysis results.
 *
 * Tier levels:
 * - Free: recommend + reasons + NSGC link only (summaryText/keyFindings/specialties/letter = null)
 * - Premium: full summary with key findings and recommended specialties
 * - Pro: everything + formatted referral letter
 *
 * Ported from Source/counseling.py `generate_referral_summary()`.
 *
 * @param carrierResults - Output of analyzeCarrierRisk()
 * @param tier - Pricing tier (default: "free")
 * @param userName - Optional user name for the referral letter (pro only)
 * @returns CounselingResult with tier-appropriate fields populated
 */
export function generateReferralSummary(
  carrierResults: CarrierResult[],
  tier: Tier = 'free',
  userName?: string,
  prsResults?: Array<{ percentile: number; trait: string }>,
  pgxResults?: Array<{ actionable: boolean; drug: string }>,
): CounselingResult {
  const [recommend, reasons] = shouldRecommendCounseling(carrierResults, prsResults, pgxResults);
  const urgency = computeUrgency(carrierResults);

  const base: CounselingResult = {
    recommend,
    urgency,
    reasons,
    nsgcUrl: NSGC_URL,
    summaryText: null,
    keyFindings: null,
    recommendedSpecialties: null,
    referralLetter: null,
    upgradeMessage:
      tier === 'free'
        ? 'Upgrade to Premium for detailed counseling summaries with key findings and recommended specialties, or Pro for a personalized referral letter.'
        : tier === 'premium'
          ? 'Upgrade to Pro for a personalized genetic counseling referral letter you can share with your healthcare provider.'
          : null,
  };

  if (tier === 'free') {
    return base;
  }

  // Premium & Pro: full summary
  const keyFindings = extractKeyFindings(carrierResults);
  const specialties = inferSpecialties(carrierResults);

  const highRiskCount = carrierResults.filter((r) => r.riskLevel === 'high_risk').length;
  const carrierCount = carrierResults.filter((r) => r.riskLevel === 'carrier_detected').length;

  const today = new Date().toISOString().split('T')[0];

  let summaryText =
    `Mergenix Genetic Counseling Summary\n` +
    `Generated: ${today}\n\n` +
    `Diseases analyzed: ${carrierResults.length}\n` +
    `High-risk findings: ${highRiskCount}\n` +
    `Carrier findings: ${carrierCount}\n\n`;

  if (recommend) {
    summaryText += 'Recommendation: Genetic counseling is advised.\n';
    for (let i = 0; i < reasons.length; i++) {
      summaryText += `  ${i + 1}. ${reasons[i]}\n`;
    }
  } else {
    summaryText += 'Recommendation: No urgent findings. Routine counseling optional.\n';
  }

  base.summaryText = summaryText;
  base.keyFindings = keyFindings;
  base.recommendedSpecialties = specialties;

  // Pro: referral letter
  if (tier === 'pro') {
    base.referralLetter = formatReferralLetter(userName ?? '', keyFindings, specialties, reasons);
  }

  return base;
}

// ─── Provider Search ────────────────────────────────────────────────────────

/**
 * Filter a provider list by specialty and/or US state.
 *
 * Ported from Source/counseling.py `find_providers_by_specialty()`.
 *
 * @param providers - Array of counseling provider entries
 * @param specialty - Filter by specialty (case-insensitive). Optional.
 * @param state - Filter by US state code (2-letter, case-insensitive). Optional.
 * @returns Filtered array of matching providers
 */
export function findProvidersBySpecialty(
  providers: CounselingProviderEntry[],
  specialty?: string,
  state?: string,
): CounselingProviderEntry[] {
  let results = providers;

  if (specialty) {
    const specLower = specialty.toLowerCase();
    results = results.filter((p) => p.specialty.some((s) => s.toLowerCase() === specLower));
  }

  if (state) {
    const stateUpper = state.toUpperCase();
    results = results.filter((p) => p.state.toUpperCase() === stateUpper);
  }

  return results;
}

// ─── Internal Helpers ────────────────────────────────────────────────────────

/**
 * Extract clinically relevant key findings for the counseling summary.
 *
 * Filters carrier results to only high_risk and carrier_detected entries,
 * extracting condition, gene, risk level, parent statuses, and inheritance.
 *
 * Ported from Source/counseling.py `_extract_key_findings()`.
 *
 * @param carrierResults - Output of analyzeCarrierRisk()
 * @returns Array of key findings
 */
function extractKeyFindings(carrierResults: CarrierResult[]): CounselingKeyFinding[] {
  const findings: CounselingKeyFinding[] = [];

  for (const r of carrierResults) {
    if (r.riskLevel === 'high_risk' || r.riskLevel === 'carrier_detected') {
      findings.push({
        condition: r.condition,
        gene: r.gene,
        riskLevel: r.riskLevel,
        parentAStatus: r.parentAStatus,
        parentBStatus: r.parentBStatus,
        inheritance: r.inheritance,
      });
    }
  }

  return findings;
}

/**
 * Suggest counseling specialties based on disease categories in the findings.
 *
 * Since CarrierResult does not include a `category` field (unlike the Python
 * dict-based results), this function infers specialties by searching the
 * `condition` and `description` fields for category keywords. This is a
 * faithful adaptation: the same keyword patterns are used ("cancer"/"oncolog",
 * "cardio"/"heart", "neuro"/"brain", "pediatr"), just applied to condition
 * and description strings instead of a dedicated category field.
 *
 * Category-to-specialty mapping (from Source/counseling.py _infer_specialties):
 * - "cancer" / "oncolog" -> cancer
 * - "cardio" / "heart" -> cardiovascular
 * - "neuro" / "brain" -> neurogenetics
 * - "pediatr" -> pediatric
 * - carrier_detected results -> carrier_screening
 * - high_risk results -> prenatal
 * - Any findings at all -> general
 *
 * Ported from Source/counseling.py `_infer_specialties()`.
 *
 * @param carrierResults - Output of analyzeCarrierRisk()
 * @returns Sorted array of recommended specialty strings
 */
function inferSpecialties(carrierResults: CarrierResult[]): CounselorSpecialty[] {
  const specs = new Set<CounselorSpecialty>();

  for (const r of carrierResults) {
    if (r.riskLevel !== 'high_risk' && r.riskLevel !== 'carrier_detected') {
      continue;
    }

    // Build a searchable text from condition and description, which together
    // contain the same keywords that the Python code found in the category field
    // (e.g., "Cancer Predisposition" category -> condition often contains "cancer")
    const searchText = `${r.condition} ${r.description}`.toLowerCase();

    // Map disease keywords to counseling specialties
    if (searchText.includes('cancer') || searchText.includes('oncolog')) {
      specs.add('cancer');
    }
    if (searchText.includes('cardio') || searchText.includes('heart')) {
      specs.add('cardiovascular');
    }
    if (searchText.includes('neuro') || searchText.includes('brain')) {
      specs.add('neurogenetics');
    }
    if (searchText.includes('pediatr')) {
      specs.add('pediatric');
    }

    // Carrier screening is always relevant for carrier results
    if (
      r.riskLevel === 'carrier_detected' ||
      (r.parentAStatus === 'carrier' && r.parentBStatus === 'carrier')
    ) {
      specs.add('carrier_screening');
    }

    // Prenatal is recommended for any high-risk finding
    if (r.riskLevel === 'high_risk') {
      specs.add('prenatal');
    }
  }

  // Always include general if we have any findings
  if (specs.size > 0) {
    specs.add('general');
  }

  return [...specs].sort();
}

/**
 * Generate a formatted referral letter for pro-tier users.
 *
 * Includes date, patient name, reasons, key findings, recommended specialties,
 * disclaimer, and NSGC link.
 *
 * Ported from Source/counseling.py `_format_referral_letter()`.
 *
 * @param userName - Patient display name
 * @param keyFindings - Key findings from the analysis
 * @param specialties - Recommended specialties
 * @param reasons - Counseling recommendation reasons
 * @returns Formatted letter string
 */
function formatReferralLetter(
  userName: string,
  keyFindings: CounselingKeyFinding[],
  specialties: CounselorSpecialty[],
  reasons: string[],
): string {
  const today = new Date().toISOString().split('T')[0];
  const nameLine = userName ? `Patient: ${userName}` : '';

  const lines: string[] = [
    'GENETIC COUNSELING REFERRAL',
    '='.repeat(40),
    `Date: ${today}`,
    nameLine,
    '',
    'Dear Genetic Counselor,',
    '',
    'This referral is generated by the Mergenix genetic analysis platform.',
    'The following findings warrant professional genetic counseling:',
    '',
  ];

  if (reasons.length > 0) {
    for (let i = 0; i < reasons.length; i++) {
      lines.push(`  ${i + 1}. ${reasons[i]}`);
    }
    lines.push('');
  }

  if (keyFindings.length > 0) {
    lines.push('Key Findings:');
    for (const f of keyFindings) {
      lines.push(
        `  - ${f.condition} (${f.gene}): ` +
          `Parent A=${f.parentAStatus}, Parent B=${f.parentBStatus} ` +
          `[${f.inheritance}]`,
      );
    }
    lines.push('');
  }

  if (specialties.length > 0) {
    lines.push(`Recommended specialties: ${specialties.join(', ')}`);
    lines.push('');
  }

  lines.push(
    'This report is for informational purposes only and does not constitute',
    'a medical diagnosis. Please review the raw data and clinical context.',
    '',
    'Sincerely,',
    'Mergenix Genetic Analysis Platform',
    `NSGC Directory: ${NSGC_URL}`,
  );

  return lines.join('\n');
}
